src/ekat/std_meta/ekat_std_map_key_iterator.hpp

```diff
@@ -0,0 +1,54 @@
+#ifndef EKAT_MAP_KEY_ITERATOR_HPP
+#define EKAT_MAP_KEY_ITERATOR_HPP
+
+#include <map>
+
+namespace ekat {
+
+/*
+ * An iterator that allows to iterate over the keys of a map only.
+ *
+ * The reason to have this class is that we can expose a way to
+ * access the keys of a map without exposing the actual content
+ * of the pair. This is very handy for map<K,std::shared_ptr<V>>,
+ * where even if the map is const, iterating over it allows one
+ * to modify the pointees.
+ * By offering a key-iterator, one can iterate over the keys of
+ * the map, without exposing the values in any way.
+ * Note: this is similar to python's dictionary 'keys()' method.
+ */
+
+template<typename MapType>
+class map_key_iterator;
+template<typename MapType>
+class map_key_const_iterator;
+
+template<typename Key, typename Value>
+class map_key_iterator<std::map<Key,Value>> final : public std::map<Key, Value>::iterator
+{
+public:
+  using map_iterator = typename std::map<Key,Value>::iterator;
+
+  map_key_iterator ( ) : map_iterator ( ) { };
+  map_key_iterator ( map_iterator it_ ) : map_iterator ( it_ ) { };
+
+  Key *operator -> ( ) { return ( Key * ) &( map_iterator::operator -> ( )->first ); }
+  Key operator * ( ) { return map_iterator::operator * ( ).first; }
+};
+
+template<typename Key, typename Value>
+class map_key_const_iterator<std::map<Key,Value>> final : public std::map<Key, Value>::const_iterator
```

Comment:
This is probably too nitpicky, but: do you want to `public`ly inherit from `std::map<K,V>::const_iterator` if you're trying not to expose the value content? Not a big deal. :-P

Reply:
Well, the class overloads `op*` and `op->`, which effectively 'hides' the base class operators, no? So one should not be able to get the std::pair stored in the map from this iterator... Unless they do `std::map<K,V>::const_iterator it = the_map_key_const_iter`. No?

Reply:
And even if I inherit privately, they can always do the aforementioned hack, I think.

Reply:
All true. Looks good to me.